tests/test_js_spidermonkey.py

```diff
@@ -40,52 +37,84 @@ def test_scripting_snippet():
     assert err.failed()
 
 
-class MockSubprocess(object):
-    "A class to mock subprocess"
+def test_unicode_escapes():
+    """Test that unicode/hex escapes are passed through to Spidermonkey."""
 
-    PIPE = True
+    result = _do_test_raw("var foo = '\\u263a\\u0027\\x27'")
 
-    def Popen(self, command, shell, stderr, stdout, stdin):
-        return MockSubprocessObject()
+    assert not result.failed()
 
+    eq_(_get_var(result, "foo"), u"\u263a''")
 
-class MockSubprocessObject(object):
-    "A class to mock a subprocess object and implement the communicate method"
 
-    def communicate(self, data):
-        data = json.dumps({"error": True,
-                           "error_message": "ReferenceError: Reflect is not defined",
-                           "line_number": 0})
-        return data, ""
+@mock.patch("subprocess.Popen")
+def test_cleanup(Popen):
+    """Test that the spidermonkey processes are terminated during cleanup."""
 
+    # Make sure we don't already have a cached shell.
+    JSShell.cleanup()
+    eq_(JSShell.shells, {})
+    assert not Popen.called
 
-@mock.patch("validator.testcases.javascript.spidermonkey.subprocess",
-            MockSubprocess())
-def test_reflectparse_presence():
-    "Tests that when Spidermonkey is too old, a proper error is produced"
+    process = mock.Mock()
+    process.stdout.readline.return_value = "{}\n"
+    Popen.return_value = process
 
-    try:
-        spidermonkey._get_tree("foo bar", "[path]")
-    except RuntimeError as err:
-        assert str(err).startswith("Spidermonkey version too old"), err
+    get_tree("hello();", shell=SPIDERMONKEY_INSTALLATION)
 
+    eq_(Popen.call_count, 1)
+    eq_(JSShell.shells.keys(), [SPIDERMONKEY_INSTALLATION])
+    assert not process.terminate.called
 
-def test_compiletime_errors():
-    "Tests that compile time errors don't break the validator"
+    decorator.cleanup()
 
-    # Syntax error
-    assert _do_test_raw("var x =;").failed()
+    assert process.terminate.called
+    eq_(JSShell.shells, {})
 
-    # Reference error
-    assert _do_test_raw("x - y = 4;").failed()
 
+@mock.patch("subprocess.Popen")
+def test_cleanup_on_error(Popen):
+    """Test that the cached shells are removed on IO error."""
 
-def test_unicode_escapes():
-    """Tests that unicode/hex escapes are passed through to Spidermonkey."""
+    # Make sure we don't already have a cached shell.
+    JSShell.cleanup()
+    eq_(JSShell.shells, {})
 
-    result = _do_test_raw("var foo = '\\u263a\\u0027\\x27'")
+    process = mock.Mock()
+    process.stdout.readline.return_value = "{}"
+    Popen.side_effect = lambda *args, **kw: process
```

Comment:
Because this test assigns a new mock to `process`. I originally had it as a `return_value`, but it wound up too messy.